feat(code_executors): Add GkeCodeExecutor for sandboxed code execution on GKE

syangx39 opened the pull request:
### Summary

This PR introduces `GkeCodeExecutor`,  a new code executor that provides a secure and scalable method for running LLM-generated code by leveraging GKE Sandbox. It serves as a robust alternative to local or standard containerized executors by leveraging the **GKE Sandbox** environment, which uses gVisor for workload isolation.

For each code execution request, it dynamically creates an ephemeral Kubernetes Job with a hardened Pod configuration, offering significant security benefits and ensuring that each code execution runs in a clean, isolated environment.

### Key Features of GkeCodeExecutor

* **Dynamic Job Creation**: Uses the Kubernetes `batch/v1` API to create a new Job for each code snippet.
* **Secure Code Mounting**: Injects code into the Pod via a temporary `ConfigMap`, which is mounted to a read-only file.
* **gVisor Sandboxing**: Enforces execution within a `gvisor` runtime for kernel-level isolation.
* **Hardened Security Context**: Pods run as non-root with all Linux capabilities dropped and a read-only root filesystem.
* **Resource Management**: Applies configurable CPU and memory limits to prevent abuse.
* **Automatic Cleanup**: Uses the `ttl_seconds_after_finished` feature on Jobs for robust, automatic garbage collection of completed Pods and Jobs.
* **Node Scheduling**: The executor uses Kubernetes `tolerations` in its Pod specification. This allows the k8s scheduler to place the execution Pod onto a **_pre-configured_** gVisor-enabled node.
* **Module Integration**: The `GkeCodeExecutor` is registered in the `code_executors/__init__.py`, making it available for use by agents. The `ImportError` handling is configured to check for the required `kubernetes` SDK.

### Execution Flow:
![54zSPHsQTMaBN3m (1)](https://github.com/user-attachments/assets/e802955d-2043-4289-8def-15b88362692b)

1. Agent invokes `GkeCodeExecutor` with the LLM-generated code. 
2. The `GkeCodeExecutor` will `execute_code` – creates a temporary `ConfigMap`, and then create a k8s `Job` to run it.
3. This Job runs a standard `python:3.11-slim` container. The image is pulled once to the node and cached. The Job will mount the ConfigMap as `/app/code.py`
4. The GkeCodeExecutor will monitor the Job to completion, fetch `stdout/stderr` logs from the container, return `CodeExecutionResult` to the LlmAgent, and ensure all temp resources are deleted.
5. The calling agent formats the result and provides a final response to the user. If the result contains error, it will retry up to [error_retry_attempts](https://github.com/google/adk-python/blob/a4d432a9e62a5e759e94168463795cd0e7d7ab72/src/google/adk/code_executors/base_code_executor.py#L36) times.

syangx39 commented:
@hangfei could you assign a reviewer for my PR

eliaslevy commented on src/google/adk/code_executors/gke_code_executor.py:
```diff
@@ -0,0 +1,219 @@
+            # The Job is cleaned up by the TTL controller, and we ensure the ConfigMap is always deleted.
+            self._cleanup_configmap(configmap_name)
+
+    def _create_job_manifest(self, job_name: str, configmap_name: str) -> client.V1Job:
```
We should annotate the k8s resources, including the Job, with the invocation id from the `InvocationContext` to have some piece of data to link sandboxes to specific executions. 

There is an execution id (aka execution_session_id) in the execution context, but its only set for "stateful" executors, and its not clear what makes an executor stateful. The `stateful` is not really referenced anywhere from what I can tell.

eliaslevy commented:
@syangx39 what about blocking network access from the sandbox? Seems like we'd want to add that in. Or are you expecting the developer to create a network policy on his own an apply it to the jobs?

syangx39 commented on src/google/adk/code_executors/gke_code_executor.py:
```diff
@@ -0,0 +1,219 @@
+        )
+        self._core_v1.create_namespaced_config_map(namespace=self.namespace, body=body)
+
+    def _cleanup_configmap(self, name: str) -> None:
```
Done. the ConfigMap is now a "child" of the Job.

syangx39 reviewed (commented):
**Two Action Items left afterwards**
1. **Follow-up PR on input/output files support with _create_pvc().**
2. regarding this comment "_@syangx39 what about blocking network access from the sandbox? Seems like we'd want to add that in. Or are you expecting the developer to create a network policy on his own an apply it to the jobs?_" \
**gVisor's default network sandboxing** in GKE specifically blocks access to sensitive host-level endpoints like the GKE metadata server. A pod running in the gVisor sandbox can still make network calls to public services on the internet. \
If gke_code_executor blocks network access, the sandboxed code will not be able to `pip install` packages from the internet. So I'm thinking -  \
The best way to handle dependencies is ahead of time. A developer would build an image with the required packages already installed via a requirements.txt file. They would then push this image to Artifact Registry and configure to use their custom image instead of the default `python:3.11`.  \
Anyway, **I will add a block_network_access flag for the flexibility and default to `True`.** But for readbility, will do it in a follow-up PR.

syangx39 commented:
@hangfei 
1. "Please add a sample under contributing/samples"
added
2. "Please add unit tests."
added 
```
tests/unittests/code_executors/test_gke_code_executor.py::TestGkeCodeExecutor::test_create_job_manifest_structure[VERTEX] PASSED                                                            [100%]
```

3. "What's the startup time for GkeCodeExecutor?"
We haven't really tested this but I can give you a ballpark figure
Here's the breakdown of startup time:
- API latency: agent making API call to k8s control plane to create ConfigMap and Job. This is negligible (miliseconds).
- Pod Scheduling: k8s scheduler assign new Pod to a gVisor-enabled node. If healthy cluster w available nodes, this is miliseconds to seconds.
- Pod Initialization: 
	- image pull is the biggest factor. The node must have the container image. For the first time (cold start), it download from registry, since python:3.11-slim is a lightweight image, it takes seconds. For the following times (cache), it's neglible.
	- gVisor Sandbox creation: slight overhead (miliseconds?)
- From my previous experience, it usually takes a few seconds (for the 1st time, the following times are instant due to cache) on a healthy cluster. We're working on cold start optimization.

4. "Should the user starts it first or it can be triggered on-demand?"
triggered on-demand, [code_executor.execute_code(...)](https://github.com/google/adk-python/blob/a858d79b3af1f42e4675b352626dfeef507a48aa/src/google/adk/flows/llm_flows/_code_execution.py#L230
) automatically run analysis code on any new data files the user provides.

5. "please create a issue and doc to update docs ad adk-docs repo. Please make sure you add a corresponding docs in adk-docs before submission."
Sure. Do you have a specific page you want me to put? 
maybe https://google.github.io/adk-docs/tools/built-in-tools/#code-execution?

hangfei commented:
> 5\. maybe https://google.github.io/adk-docs/tools/built-in-tools/#code-execution?

this works: maybe https://google.github.io/adk-docs/tools/built-in-tools/#code-execution?

hangfei commented on src/google/adk/code_executors/gke_code_executor.py:
```diff
@@ -0,0 +1,248 @@
+    each execution request. The user's code is mounted via a ConfigMap, and the
+    Pod is hardened with a strict security context and resource limits.
+
+    Key Features:
```
fair point.

hangfei commented on src/google/adk/code_executors/gke_code_executor.py:
```diff
@@ -0,0 +1,248 @@
+    with specific RBAC permissions to function. The agent's pod needs permissions
+    to create/watch Jobs, create/delete ConfigMaps, and list Pods to read logs.
+    For a complete, working example of the required Role and RoleBinding, see the
+    file at: contributing/samples/gke_agent_sandbox/deployment_rbac.yaml
```
that's better. thanks.

hangfei commented on src/google/adk/code_executors/gke_code_executor.py:
```diff
@@ -0,0 +1,248 @@
+    file at: contributing/samples/gke_agent_sandbox/deployment_rbac.yaml
+    """
+    namespace: str = "default"
+    image: str = "python:3.11-slim"
```
please. thanks.

hangfei commented:
Overall LGTM. Thanks!

Let's do the following before merge: 
1. have a adk-docs PR for this change.
2. fix the tests.

syangx39 commented:
@hangfei 
The doc PR got merged. Please merge this PR as well for consistency.
https://google.github.io/adk-docs/tools/built-in-tools/#gke-code-executor

hangfei commented:
thanks. 

1. there are some formatting and tests failures. plz fix. 

2. could you share the steps to test the code execuctor so we can test it?

seanzhou1023 commented on contributing/samples/gke_agent_sandbox/deployment_rbac.yaml:
```diff
@@ -0,0 +1,50 @@
+apiVersion: v1
```
great, could you add a README.md to this sample agent and putting information above in it , so that users understand the usage of this file ?

seanzhou1023 commented on src/google/adk/code_executors/__init__.py:
```diff
@@ -49,3 +49,14 @@
       ' warning.'
   )
+
+try:
```
the original codes in this file is problematic, we've just fixed it, let's follow the latest style.

syangx39 commented:
> 1. there are some formatting and tests failures. plz fix.
> 2. could you share the steps to test the code execuctor so we can test it?

1. sounds good. I forgot to add "kubernetes" in [test] in pyproject.toml, fixed it
2. Sounds good. This is the plan:

### GkeCodeExecutor Validation Plan

#### Step 1: Set Up the GKE Autopilot Cluster
```
gcloud container clusters create-auto adk-sandbox-cluster \
  --project="<YOUR_PROJECT_ID>" \
  --region="<YOUR_GCP_REGION>"
```

#### Step 2: Deploy the Agent and Send Predictions
1. Apply RBAC Permissions `kubectl apply -f deployment_rbac.yaml`.
2. Build and Push the adk Agent Image 
3. Update `deployment_agent.yaml` with your image path and run `kubectl apply -f ...`
4. `kubectl port-forward svc/adk-agent-service 8080:80 -n adk-sandbox`
5. gVisor Sandbox Execution
```
curl -X POST http://localhost:8080/invoke \
-H "Content-Type: application/json" \
-d '{
    "prompt": "Write a Python script to query the GKE metadata server for the node'\''s instance ID and print the result. If it fails, print the error."
}'
```
